Add FrameProcessor.is_usable and error categories, and stop retrying services that can't recover (#5242)
pipecat-ai/pipecat: changelog/5242.changed.3.md

> - `ServiceSwitcher` now fails over on the errors that cost it a service, and answers for its services in the errors it reports.

Reviewer: Confusing wording

Reviewer: Make sure to mention `is_usable` here, for clarity, too

>     `ServiceSwitcherStrategyFailover` switches once the active service reports an error that leaves it unable to do its job, rather than on any error, so a provider hiccup no longer costs a failover. It switches to the next service that can still be given work, passing over any that can't, and a successful switch consumes the error: the switcher went on doing its job, so nothing upstream needs to act on it.

Reviewer: ```suggestion
    `ServiceSwitcherStrategyFailover` switches only once the active service reports an error that leaves it unable to do its job, rather than on any error, so a provider hiccup no longer costs a failover. It switches to the next service that is still usable, and a successful switch consumes the error: the switcher went on doing its job, so nothing upstream needs to act on it.
```

>     Errors from a service the switcher isn't using go no further, since nothing about a service held in reserve bears on whether the switcher can do its job — watch that service's own `on_usable_changed` to hear about it. Without a switch, the error is re-reported against the switcher itself, naming the service that failed. The pipeline deals with the switcher rather than with the services inside it, so whether the failure matters depends on what the switcher has left: it reports itself unusable only once none of its services can work, and a service whose configuration was rejected never writes off the switcher along with it. Bringing any service back with `set_usable(True)` brings the switcher back with it. The switcher raises `on_usable_changed` for itself when that reading moves, so watching the switcher is enough to hear about the services inside it.

Reviewer: Nice—so `on_usable_changed` works on the switcher as expected, with it reporting the switcher unusable when its last usable service becomes unusable, then reporting the switcher usable again when one of its services becomes usable again (reflecting the switcher's own `is_usable`)?

Author: Exactly, `on_usable_changed` is always reflecting the switcher's own `is_usable`.
